finetune/input_pipeline.py

```diff
@@ -0,0 +1,209 @@
+            map(lambda xy: self.text_to_tokens_mask(*xy), dataset()))
+        shape_def = self.feed_shape_type_def()
+        if len(self.config.visible_gpus) > 1:
+            dataset_encoded_ = dataset_encoded
+            dataset_encoded = lambda: map(self._adapt_feats_for_distributed, dataset_encoded_())
+            shape_def = [self._adapt_feats_for_distributed(s) for s in shape_def]
+        return Dataset.from_generator(dataset_encoded, *shape_def)
+
+    def _dataset_without_targets(self, Xs):
+        if not callable(Xs):
+            Xs_fn = lambda: Xs
+        else:
+            Xs_fn = Xs
+
+        dataset_encoded = lambda: itertools.chain.from_iterable(map(self.text_to_tokens_mask, Xs_fn()))
+        types, shapes = self.feed_shape_type_def()
+        return Dataset.from_generator(dataset_encoded, types[0], shapes[0])  # 0s cut out the targets
+
+    def _get_train_input_fns(self, Xs, Y=None, batch_size=None, val_size=None):
+        batch_size = batch_size or self.config.batch_size
+
+        shuffle_buffer_size = 100
+        val_size = val_size or 0
+        prefetch_buffer = 2  # breaks the pipeline to allow concurrency
+        if Y is not None:
+            self._post_data_initialization(Y)
+            dataset = lambda: self._dataset_with_targets(Xs, Y)
+        else:
+            dataset = lambda: self._dataset_without_targets(Xs)
+
+        val_dataset = lambda: dataset().shuffle(shuffle_buffer_size, seed=self.config.seed).take(
+            val_size).batch(batch_size).prefetch(prefetch_buffer)
+        train_dataset = lambda: dataset().shuffle(shuffle_buffer_size, seed=self.config.seed).skip(
+            val_size).batch(batch_size).repeat(self.config.n_epochs).prefetch(prefetch_buffer)
+
+        return val_dataset, train_dataset
+
+    def _get_predict_input_fn(self, Xs, batch_size=None):
+        batch_size = batch_size or self.config.batch_size
+        prefetch_buffer = 2  # breaks the pipeline to allow concurrency
+        tf_dataset = lambda: self._dataset_without_targets(Xs)
+        return lambda: tf_dataset().batch(batch_size).prefetch(prefetch_buffer)
+
+    @property
+    def pad_idx(self):
+        if self.pad_idx_ is None:
+            self.pad_idx_ = list(self.label_encoder.classes_).index(self.config.pad_token)
+        return self.pad_idx_
+
+    def _format_for_encoding(self, X):
+        """
+        Most subclasses take in inputs as:
+            List (batch) of list (docs)
+
+        Encode_multi_input expect the following format:
+            List (batch) of list (docs) of list (subseqs) of text
+
+        This method is responsible for standardizing inputs to the above format
+        """
+        return [[X]]
+
+    def _text_to_ids(self, Xs, Y=None, pad_token=PAD_TOKEN):
+        Xs = self._format_for_encoding(Xs)
+        if self.config.chunk_long_sequences and len(Xs) == 1:
+            # can only chunk single sequence inputs
+            chunk_size = self.config.max_length - 2
```

Comment:
Maybe worth breaking this out into a separate routine -- (yes, I know I wrote this, it's more of a note to myself).